agent/exec/controller.go

```diff
@@ -158,6 +158,7 @@ func Do(ctx context.Context, task *api.Task, ctlr Controller) (*api.TaskStatus,
 		current := status.State
 		status.State = state
 		status.Message = msg
+		status.AppliedAt = ptypes.MustTimestampProto(time.Now())
```

Comment:
This isn't how `AppliedAt` was intended to be used.

```go
	// AppliedAt gives a timestamp of when this status update was applied to
	// the Task object.
```

The status update gets applied to the `Task` object in the dispatcher, but this code is in the agent.

I think you should use `time.Now()` in the dispatcher instead of relying on a timestamp from the `TaskStatus`. You can't compare a timestamp that comes from a worker with one that was originally recorded on the manager, because their clocks may be out of sync. Even if there's only a few seconds difference, it would wildly throw off the results.

Reply:
>This isn't how AppliedAt was intended to be used. The status update gets applied to the Task object in the dispatcher, but this code is in the agent.

You mean it gets applied/persisted to the task object in the store ?

The actual status change happens in different components as the task passes through them to the worker.
I agree with what you're saying, but that will not give us accurate results either since the actual time of status change is when it actually happens.

Please correct me if I'm wrong.

Reply:
> You mean it gets applied/persisted to the task object in the store ?

Yes, that's what it's supposed to timestamp.

> I agree with what you're saying, but that will not give us accurate results either since the actual time of status change is when it actually happens.
>
> Please correct me if I'm wrong.

You're correct, neither approach will be completely accurate. However, I think counting the extra network latency is less catastrophic than assuming perfectly sync'd clocks.

Reply:
>I think counting the extra network latency is less catastrophic than assuming perfectly sync'd clocks.

I see the point. Assuming perfectly synced clocks may give ambiguous or wrong results, but including network latency will include the extra network latency but the results will not be wrong.